apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/StorageClientImpl.UploadObject.cs

```diff
@@ -92,41 +93,113 @@ public sealed partial class StorageClientImpl : StorageClient
+                if (progress != null)
+                {
+                    _mediaUpload.ProgressChanged += progress.Report;
+                }
+
+                var validationMode = options?.UploadValidationMode ?? UploadObjectOptions.DefaultValidationMode;
+                GaxPreconditions.CheckEnumValue(validationMode, nameof(UploadObjectOptions.UploadValidationMode));
+                switch (validationMode)
+                {
+                    case UploadValidationMode.DeleteAndThrow:
+                        _crc = new Crc32c();
+                        _mediaUpload.UploadStreamInterceptor += _crc.UpdateHash;
+                        _validationFailureAction = obj => client.DeleteObject(obj, new DeleteObjectOptions { Generation = obj.Generation });
+                        _validationFailureAsyncAction = (obj, token) => client.DeleteObjectAsync(obj, new DeleteObjectOptions { Generation = obj.Generation }, token);
+                        break;
+                    case UploadValidationMode.ThrowOnly:
+                        _crc = new Crc32c();
+                        _mediaUpload.UploadStreamInterceptor += _crc.UpdateHash;
+                        break;
+                }
             }
-            var finalProgress = await mediaUpload.UploadAsync(cancellationToken).ConfigureAwait(false);
-            if (finalProgress.Exception != null)
+
+            internal Object Execute()
+            {
+                _mediaUpload.Upload();
+                CheckFinalProgress();
+                var result = _mediaUpload.ResponseBody;
+                var hash = _crc == null ? result.Crc32c : Convert.ToBase64String(_crc.GetHash());
+                if (hash != result.Crc32c)
+                {
+                    AggregateException additionalFailures = null;
+                    try
+                    {
+                        _validationFailureAction.Invoke(result);
+                    }
+                    catch (Exception e)
+                    {
+                        additionalFailures = new AggregateException(e);
+                    }
+                    throw new UploadValidationException(hash, result, additionalFailures);
+                }
+                return result;
+            }
+
+            internal async Task<Object> ExecuteAsync(CancellationToken cancellationToken)
+            {
+                await _mediaUpload.UploadAsync(cancellationToken).ConfigureAwait(false);
+                CheckFinalProgress();
+                var result = _mediaUpload.ResponseBody;
+                var hash = _crc == null ? result.Crc32c : Convert.ToBase64String(_crc.GetHash());
+                if (hash != result.Crc32c)
+                {
+                    AggregateException additionalFailures = null;
+                    try
+                    {
+                        var task = _validationFailureAsyncAction?.Invoke(result, cancellationToken) ?? Task.FromResult(0);
```

Comment:
Just using an `if` statement might be cleaner/clearer here?

Reply:
Will have a try in a few ways.